presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java

```diff
@@ -386,6 +438,239 @@ else if (exchange.getType() == ExchangeNode.Type.REPARTITION) {
+            PartitioningMetadata partitioningMetadata = new PartitioningMetadata(partitioningHandle, partitionColumns);
+
+            TableHandle temporaryTable = metadata.createTemporaryTable(
+                    session,
+                    connectorId.getCatalogName(),
+                    ImmutableList.copyOf(temporaryTableColumns.values()),
+                    Optional.of(partitioningMetadata));
+
+            TableScanNode scan = createTemporaryTableScan(
+                    temporaryTable,
+                    exchange.getOutputSymbols(),
+                    temporaryTableColumns,
+                    partitioningMetadata);
+
+            checkArgument(
+                    !exchange.getPartitioningScheme().isReplicateNullsAndAny(),
+                    "materialized remote exchange is not supported when replicateNullsAndAny is needed");
+            TableFinishNode write = createTemporaryTableWrite(
+                    temporaryTable,
+                    temporaryTableColumns,
+                    exchange.getOutputSymbols(),
+                    exchange.getInputs(),
+                    exchange.getSources(),
+                    constantExpressions,
+                    partitioningMetadata);
+
+            FragmentProperties writeProperties = new FragmentProperties(new PartitioningScheme(
+                    Partitioning.create(SINGLE_DISTRIBUTION, ImmutableList.of()),
+                    write.getOutputSymbols()));
+            writeProperties.setCoordinatorOnlyDistribution();
+
+            SubPlan childFragmentPlan = buildSubPlan(write, writeProperties, context);
+            context.get().addChildren(ImmutableList.of(childFragmentPlan));
+            context.get().addDependency(childFragmentPlan.getFragment().getId());
+
+            return visitTableScan(scan, context);
+        }
+
+        private List<Symbol> assignPartitioningSymbols(Partitioning partitioning, Map<Symbol, Expression> constantPartitioningSymbols)
+        {
+            ImmutableList.Builder<Symbol> result = ImmutableList.builder();
+            for (ArgumentBinding argumentBinding : partitioning.getArguments()) {
+                Symbol symbol;
+                if (argumentBinding.isConstant()) {
+                    NullableValue constant = argumentBinding.getConstant();
+                    Expression expression = literalEncoder.toExpression(constant.getValue(), constant.getType());
+                    symbol = symbolAllocator.newSymbol(expression, constant.getType());
+                    constantPartitioningSymbols.put(symbol, expression);
+                }
+                else {
+                    symbol = argumentBinding.getSymbol();
+                }
+                result.add(symbol);
+            }
+            return result.build();
+        }
+
+        private Map<Symbol, ColumnMetadata> assignColumnNames(Collection<Symbol> outputSymbols, Collection<Symbol> constantPartitioningSymbols)
```

Comment:
I would call it `assignTemporaryTableColumnNames` or something similar than the more general name here.